File: sdks/java/src/main/java/com/legacybaas/LegacyBaaSClient.java
package com.legacybaas;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.datatype.jsr310.JavaTimeModule;
import com.legacybaas.config.ClientConfig;
import com.legacybaas.config.Environment;
import com.legacybaas.exceptions.LegacyBaaSException;
import com.legacybaas.exceptions.AuthenticationException;
import com.legacybaas.exceptions.ValidationException;
import com.legacybaas.exceptions.NetworkException;
import com.legacybaas.exceptions.RateLimitException;
import com.legacybaas.exceptions.ServiceUnavailableException;
import com.legacybaas.http.HttpClient;
import com.legacybaas.models.common.ApiResponse;
import com.legacybaas.models.common.HealthStatus;
import com.legacybaas.services.SwiftProcessor;
import com.legacybaas.services.BlockchainRouter;
import com.legacybaas.services.BancsIntegration;
import com.legacybaas.services.AnalyticsService;
import com.legacybaas.services.WebhookHandler;
import com.legacybaas.utils.ApiKeyValidator;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.Closeable;
import java.time.Duration;
import java.util.Map;
import java.util.concurrent.CompletableFuture;

/**
 * Main Legacy B2BaaS Platform Client for Java
 * 
 * <p>Provides access to all platform services including SWIFT processing,
 * blockchain routing, BaNCS integration, and analytics.</p>
 * 
 * <p>Example usage:</p>
 * <pre>{@code
 * LegacyBaaSClient client = LegacyBaaSClient.builder()
 *     .apiKey("your-api-key")
 *     .environment(Environment.PRODUCTION)
 *     .build();
 * 
 * // Check platform health
 * HealthStatus health = client.health().get();
 * System.out.println("Platform status: " + health.getStatus());
 * 
 * // Process SWIFT MT103 message
 * MT103Message mt103 = MT103Message.builder()
 *     .transactionReference("TXN123456")
 *     .valueDate(LocalDate.now())
 *     .currency("USD")
 *     .amount(BigDecimal.valueOf(10000))
 *     .orderingCustomer(Customer.builder()
 *         .account("123456789")
 *         .name("John Doe")
 *         .build())
 *     .beneficiaryCustomer(Customer.builder()
 *         .account("987654321")
 *         .name("Jane Smith")
 *         .build())
 *     .build();
 * 
 * SwiftProcessingResult result = client.swift().processMT103(mt103).get();
 * System.out.println("Transaction ID: " + result.getTransactionId());
 * 
 * // Close client when done
 * client.close();
 * }</pre>
 * 
 * @author Legacy B2BaaS Platform
 * @version 1.0.0
 * @since 1.0.0
 */
public class LegacyBaaSClient implements Closeable {
    
    private static final Logger logger = LoggerFactory.getLogger(LegacyBaaSClient.class);
    private static final String SDK_VERSION = "1.0.0";
    
    private final String apiKey;
    private final ClientConfig config;
    private final HttpClient httpClient;
    private final ObjectMapper objectMapper;
    
    // Service instances
    private final SwiftProcessor swiftProcessor;
    private final BlockchainRouter blockchainRouter;
    private final BancsIntegration bancsIntegration;
    private final AnalyticsService analyticsService;
    private final WebhookHandler webhookHandler;
    
    /**
     * Constructs a new Legacy B2BaaS client
     * 
     * @param apiKey Your Legacy B2BaaS API key
     * @param config Client configuration
     * @throws AuthenticationException if API key is invalid
     * @throws IllegalArgumentException if required parameters are null
     */
    public LegacyBaaSClient(@NotNull String apiKey, @NotNull ClientConfig config) {
        if (apiKey == null || apiKey.trim().isEmpty()) {
            throw new IllegalArgumentException("API key cannot be null or empty");
        }
        if (config == null) {
            throw new IllegalArgumentException("Client config cannot be null");
        }
        
        // Validate API key format
        if (!ApiKeyValidator.isValid(apiKey)) {
            throw new AuthenticationException("Invalid API key format");
        }
        
        this.apiKey = apiKey;
        this.config = config;
        
        // Initialize object mapper
        this.objectMapper = new ObjectMapper()
            .registerModule(new JavaTimeModule())
            .findAndRegisterModules();
        
        // Initialize HTTP client
        this.httpClient = new HttpClient(apiKey, config, objectMapper);
        
        // Initialize services
        this.swiftProcessor = new SwiftProcessor(httpClient);
        this.blockchainRouter = new BlockchainRouter(httpClient);
        this.bancsIntegration = new BancsIntegration(httpClient);
        this.analyticsService = new AnalyticsService(httpClient);
        this.webhookHandler = new WebhookHandler(httpClient);
        
        logger.info("Legacy B2BaaS Client initialized for environment: {}", config.getEnvironment());
    }
    
    /**
     * Creates a new builder for configuring the client
     * 
     * @return A new client builder
     */
    public static Builder builder() {
        return new Builder();
    }
    
    /**
     * Get SWIFT processing service
     * 
     * @return SWIFT processor instance
     */
    public SwiftProcessor swift() {
        return swiftProcessor;
    }
    
    /**
     * Get blockchain routing service
     * 
     * @return Blockchain router instance
     */
    public BlockchainRouter blockchain() {
        return blockchainRouter;
    }
    
    /**
     * Get BaNCS integration service
     * 
     * @return BaNCS integration instance
     */
    public BancsIntegration bancs() {
        return bancsIntegration;
    }
    
    /**
     * Get analytics service
     * 
     * @return Analytics service instance
     */
    public AnalyticsService analytics() {
        return analyticsService;
    }
    
    /**
     * Get webhook handler service
     * 
     * @return Webhook handler instance
     */
    public WebhookHandler webhooks() {
        return webhookHandler;
    }
    
    /**
     * Check platform health status
     * 
     * @return Future containing health status
     */
    public CompletableFuture<HealthStatus> health() {
        return httpClient.get("/health", HealthStatus.class);
    }
    
    /**
     * Test API connectivity
     * 
     * @return Future containing ping response
     */
    public CompletableFuture<Map<String, Object>> ping() {
        return httpClient.get("/ping", Map.class);
    }
    
    /**
     * Get account information
     * 
     * @return Future containing account details
     */
    public CompletableFuture<Map<String, Object>> getAccount() {
        return httpClient.get("/account", Map.class);
    }
    
    /**
     * Get API usage statistics
     * 
     * @param period Time period ("day", "week", "month")
     * @return Future containing usage statistics
     */
    public CompletableFuture<Map<String, Object>> getUsage(@NotNull String period) {
        return httpClient.get("/usage?period=" + period, Map.class);
    }
    
    /**
     * Get API usage statistics for the current day
     * 
     * @return Future containing usage statistics
     */
    public CompletableFuture<Map<String, Object>> getUsage() {
        return getUsage("day");
    }
    
    /**
     * Get client configuration
     * 
     * @return Client configuration
     */
    public ClientConfig getConfig() {
        return config;
    }
    
    /**
     * Get SDK version
     * 
     * @return SDK version string
     */
    public static String getSdkVersion() {
        return SDK_VERSION;
    }
    
    /**
     * Close the client and free resources
     */
    @Override
    public void close() {
        try {
            if (httpClient != null) {
                httpClient.close();
            }
            logger.info("Legacy B2BaaS Client closed");
        } catch (Exception e) {
            logger.warn("Error closing client: {}", e.getMessage());
        }
    }
    
    /**
     * Builder for creating LegacyBaaSClient instances
     */
    public static class Builder {
        private String apiKey;
        private Environment environment = Environment.PRODUCTION;
        private String baseUrl;
        private Duration timeout = Duration.ofSeconds(30);
        private int maxRetries = 3;
        private Duration retryDelay = Duration.ofSeconds(1);
        private boolean enableLogging = true;
        
        public Builder apiKey(@NotNull String apiKey) {
            this.apiKey = apiKey;
            return this;
        }
        
        public Builder environment(@NotNull Environment environment) {
            this.environment = environment;
            return this;
        }
        
        public Builder baseUrl(@NotNull String baseUrl) {
            this.baseUrl = baseUrl;
            return this;
        }
        
        public Builder timeout(@NotNull Duration timeout) {
            this.timeout = timeout;
            return this;
        }
        
        public Builder maxRetries(int maxRetries) {
            this.maxRetries = maxRetries;
            return this;
        }
        
        public Builder retryDelay(@NotNull Duration retryDelay) {
            this.retryDelay = retryDelay;
            return this;
        }
        
        public Builder enableLogging(boolean enableLogging) {
            this.enableLogging = enableLogging;
            return this;
        }
        
        public LegacyBaaSClient build() {
            if (apiKey == null || apiKey.trim().isEmpty()) {
                throw new IllegalArgumentException("API key is required");
            }
            
            ClientConfig config = ClientConfig.builder()
                .environment(environment)
                .baseUrl(baseUrl != null ? baseUrl : environment.getBaseUrl())
                .timeout(timeout)
                .maxRetries(maxRetries)
                .retryDelay(retryDelay)
                .enableLogging(enableLogging)
                .build();
                
            return new LegacyBaaSClient(apiKey, config);
        }
    }
}

File: sdks/android/src/main/java/com/legacybaas/android/sdk/LegacyBaaSClient.kt
package com.legacybaas.android.sdk

import android.content.Context
import kotlinx.coroutines.*
import kotlinx.coroutines.flow.Flow
import kotlinx.serialization.json.Json
import okhttp3.*
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.RequestBody.Companion.toRequestBody
import timber.log.Timber
import java.io.IOException
import java.util.concurrent.TimeUnit
import com.legacybaas.android.sdk.configuration.LegacyBaaSConfiguration
import com.legacybaas.android.sdk.configuration.Environment
import com.legacybaas.android.sdk.exceptions.*
import com.legacybaas.android.sdk.models.common.*
import com.legacybaas.android.sdk.services.*
import com.legacybaas.android.sdk.utils.ApiKeyValidator

/**
 * Main Legacy B2BaaS Platform Client for Android
 *
 * Provides access to all platform services including SWIFT processing,
 * blockchain routing, BaNCS integration, and analytics.
 *
 * Example usage:
 * ```kotlin
 * val client = LegacyBaaSClient("your-api-key")
 *
 * // Check platform health
 * lifecycleScope.launch {
 *     try {
 *         val health = client.getHealth()
 *         Log.d("Health", "Platform status: ${health.status}")
 *     } catch (e: Exception) {
 *         Log.e("Error", "Health check failed: ${e.message}")
 *     }
 * }
 *
 * // Process SWIFT MT103 message
 * val mt103 = MT103Message(
 *     transactionReference = "TXN123456",
 *     currency = "USD",
 *     amount = 10000.0,
 *     orderingCustomer = Customer("123456789", "John Doe"),
 *     beneficiaryCustomer = Customer("987654321", "Jane Smith")
 * )
 *
 * lifecycleScope.launch {
 *     try {
 *         val result = client.swift.processMT103(mt103)
 *         Log.d("SWIFT", "Transaction ID: ${result.transactionId}")
 *     } catch (e: Exception) {
 *         Log.e("Error", "Transaction failed: ${e.message}")
 *     }
 * }
 * ```
 */
class LegacyBaaSClient @JvmOverloads constructor(
    private val apiKey: String,
    private val configuration: LegacyBaaSConfiguration = LegacyBaaSConfiguration.default(),
    context: Context? = null
) {
    
    companion object {
        /**
         * SDK version
         */
        const val SDK_VERSION = BuildConfig.SDK_VERSION
        
        private const val MEDIA_TYPE_JSON = "application/json; charset=utf-8"
    }
    
    private val json = Json {
        ignoreUnknownKeys = true
        isLenient = true
        encodeDefaults = false
    }
    
    private val client: OkHttpClient
    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.IO)
    
    /**
     * SWIFT processing service
     */
    val swift: SwiftService by lazy { SwiftService(this) }
    
    /**
     * Blockchain routing service
     */
    val blockchain: BlockchainService by lazy { BlockchainService(this) }
    
    /**
     * BaNCS integration service
     */
    val bancs: BancsService by lazy { BancsService(this) }
    
    /**
     * Analytics service
     */
    val analytics: AnalyticsService by lazy { AnalyticsService(this) }
    
    /**
     * Webhook handler service
     */
    val webhooks: WebhookService by lazy { WebhookService(this) }
    
    /**
     * COBOL transpiler service
     */
    val cobol: CobolTranspilerService by lazy { CobolTranspilerService(this) }
    
    init {
        require(ApiKeyValidator.isValid(apiKey)) { "Invalid API key format" }
        
        // Configure HTTP client
        client = OkHttpClient.Builder()
            .connectTimeout(configuration.timeout.toLong(), TimeUnit.SECONDS)
            .readTimeout(configuration.timeout.toLong(), TimeUnit.SECONDS)
            .writeTimeout(configuration.timeout.toLong(), TimeUnit.SECONDS)
            .addInterceptor(AuthInterceptor(apiKey))
            .addInterceptor(UserAgentInterceptor())
            .addInterceptor(RetryInterceptor(configuration.maxRetries))
            .apply {
                if (configuration.enableLogging) {
                    addInterceptor(LoggingInterceptor())
                }
            }
            .build()
        
        if (configuration.enableLogging) {
            Timber.tag("LegacyBaaS").i("Android SDK initialized for environment: ${configuration.environment}")
        }
    }
    
    /**
     * Check platform health status
     */
    suspend fun getHealth(): HealthStatus = withContext(Dispatchers.IO) {
        performRequest<HealthStatus>("/health", HttpMethod.GET)
    }
    
    /**
     * Test API connectivity
     */
    suspend fun ping(): PingResponse = withContext(Dispatchers.IO) {
        performRequest<PingResponse>("/ping", HttpMethod.GET)
    }
    
    /**
     * Get account information
     */
    suspend fun getAccount(): AccountInfo = withContext(Dispatchers.IO) {
        performRequest<AccountInfo>("/account", HttpMethod.GET)
    }
    
    /**
     * Get API usage statistics
     */
    suspend fun getUsage(period: String = "day"): UsageInfo = withContext(Dispatchers.IO) {
        performRequest<UsageInfo>("/usage?period=$period", HttpMethod.GET)
    }
    
    /**
     * Get client configuration
     */
    fun getConfiguration(): LegacyBaaSConfiguration = configuration
    
    /**
     * Close the client and cleanup resources
     */
    fun close() {
        scope.cancel()
        client.dispatcher.executorService.shutdown()
        client.connectionPool.evictAll()
        client.cache?.close()
    }
    
    // Internal methods
    
    internal suspend inline fun <reified T> performRequest(
        endpoint: String,
        method: HttpMethod,
        body: Any? = null
    ): T = withContext(Dispatchers.IO) {
        val url = buildUrl(endpoint)
        val request = buildRequest(url, method, body)
        
        try {
            val response = client.newCall(request).await()
            handleResponse<T>(response)
        } catch (e: IOException) {
            throw NetworkException("Network error: ${e.message}", e)
        }
    }
    
    private fun buildUrl(endpoint: String): String {
        val baseUrl = configuration.baseUrl ?: configuration.environment.baseUrl
        return "$baseUrl/api/v1$endpoint"
    }
    
    private fun buildRequest(url: String, method: HttpMethod, body: Any?): Request {
        val requestBuilder = Request.Builder().url(url)
        
        when (method) {
            HttpMethod.GET -> requestBuilder.get()
            HttpMethod.POST -> {
                val requestBody = body?.let {
                    json.encodeToString(kotlinx.serialization.serializer(), it)
                        .toRequestBody(MEDIA_TYPE_JSON.toMediaType())
                } ?: ByteArray(0).toRequestBody()
                requestBuilder.post(requestBody)
            }
            HttpMethod.PUT -> {
                val requestBody = body?.let {
                    json.encodeToString(kotlinx.serialization.serializer(), it)
                        .toRequestBody(MEDIA_TYPE_JSON.toMediaType())
                } ?: ByteArray(0).toRequestBody()
                requestBuilder.put(requestBody)
            }
            HttpMethod.DELETE -> requestBuilder.delete()
        }
        
        return requestBuilder.build()
    }
    
    private suspend inline fun <reified T> handleResponse(response: Response): T {
        val body = response.body?.string() ?: ""
        
        when (response.code) {
            in 200..299 -> {
                return try {
                    json.decodeFromString<T>(body)
                } catch (e: Exception) {
                    throw SerializationException("Failed to parse response: ${e.message}", e)
                }
            }
            400 -> throw ValidationException(parseErrorMessage(body))
            401 -> throw AuthenticationException(parseErrorMessage(body))
            429 -> {
                val retryAfter = response.header("Retry-After")?.toIntOrNull()
                throw RateLimitException(parseErrorMessage(body), retryAfter)
            }
            503 -> throw ServiceUnavailableException(parseErrorMessage(body))
            else -> throw HttpException(response.code, parseErrorMessage(body))
        }
    }
    
    private fun parseErrorMessage(body: String): String {
        return try {
            val errorResponse = json.decodeFromString<ErrorResponse>(body)
            errorResponse.message ?: "Unknown error"
        } catch (e: Exception) {
            "Unknown error"
        }
    }
    
    // HTTP Method enum
    internal enum class HttpMethod {
        GET, POST, PUT, DELETE
    }
}

// Extension function for async OkHttp calls
private suspend fun Call.await(): Response {
    return suspendCancellableCoroutine { continuation ->
        continuation.invokeOnCancellation {
            cancel()
        }
        
        enqueue(object : Callback {
            override fun onResponse(call: Call, response: Response) {
                continuation.resume(response)
            }
            
            override fun onFailure(call: Call, e: IOException) {
                continuation.resumeWithException(e)
            }
        })
    }
}